modAionImpl/src/org/aion/zero/impl/db/PendingBlockStore.java

```diff
@@ -0,0 +1,816 @@
+
+                @Override
+                public List<byte[]> deserialize(byte[] stream) {
+                    RLPList list = (RLPList) RLP.decode2(stream).get(0);
+                    List<byte[]> res = new ArrayList<>(list.size());
+
+                    for (RLPElement aList : list) {
+                        res.add(aList.getRLPData());
+                    }
+                    return res;
+                }
+            };
+
+    private static final Serializer<List<AionBlock>, byte[]> BLOCK_LIST_RLP_SERIALIZER =
+            new Serializer<>() {
+                @Override
+                public byte[] serialize(List<AionBlock> object) {
+                    byte[][] infoList = new byte[object.size()][];
+                    int i = 0;
+                    for (AionBlock b : object) {
+                        infoList[i] = b.getEncoded();
+                        i++;
+                    }
+                    return RLP.encodeList(infoList);
+                }
+
+                @Override
+                public List<AionBlock> deserialize(byte[] stream) {
+                    RLPList list = (RLPList) RLP.decode2(stream).get(0);
+                    List<AionBlock> res = new ArrayList<>(list.size());
+
+                    for (RLPElement aList : list) {
+                        res.add(new AionBlock(aList.getRLPData()));
+                    }
+                    return res;
+                }
+            };
+
+    /**
+     * Stores a single block in the pending block store for importing later when the chain reaches
+     * the needed height and the parent block gets imported. Is used by the functionality receiving
+     * status blocks.
+     *
+     * @implNote The status blocks received impact the functionality of the base value generation
+     *     {@link #nextBase(long, long)} for requesting blocks ahead of import time.
+     */
+    public boolean addStatusBlock(AionBlock block) {
+
+        // nothing to do with null parameter
+        if (block == null) {
+            return false;
+        }
+
+        lock.writeLock().lock();
+
+        try {
+            // skip if already stored
+            if (!indexSource.get(block.getHash()).isPresent()) {
+
+                // find parent queue hash
+                Optional<byte[]> existingQueueHash = indexSource.get(block.getParentHash());
+                byte[] currentQueueHash = null;
+                List<AionBlock> currentQueue = null;
+
+                // get existing queue if present
+                if (existingQueueHash.isPresent()) {
+                    // using parent queue hash
+                    currentQueueHash = existingQueueHash.get();
+
+                    // append block to queue
+                    currentQueue = queueSource.get(currentQueueHash);
+                } // do not add else here!
+
+                // when no queue exists OR problem with existing queue
+                if (currentQueue == null || currentQueue.size() == 0) {
```

Comment:
use "isEmpty()" instead of size() == 0

Reply:
https://stackoverflow.com/questions/1508975/why-is-list-size0-slower-than-list-isempty-in-java

Reply:
updated in 0742fff